### PR TITLE
8292995: improve the SA page cache

### PR DESCRIPTION
**Pull request by plummercj**
The page caching support in SA is woefully dated. I think it has stayed the same for over 20 years when it was originally done for solarix-x86. This code has been replicated for every port. Currently all ports only have a 16mb cache. They use 4k pages and there are 4k of them.

I think the 4k page size is fine. The following comment appears in all the ports:

            // ... This is a cache of 4096 4K pages, or 16 MB. The page
            // size must be adjusted to be the hardware's page size.
            // (FIXME: should pick this up from the debugger.)

I disagree with this. Matching the possibly very large hardware page size (I think maybe they meant OS page size) would require the SA page cache to be very very large, using a lot of java heap space. It would also require a lot of unnecessary copying from the debuggee process's memory. There's no reason for the SA cache's page size to match the OS page size.

However, 16mb seems very small. I tried 256mb and this gave about a 10% performance improvement in a heap dump, and is still fairly small, so I think it is a reasonable adjustment.

Another comment you see in all the ports (copied from solaris-x86) is:

            // FIXME: re-test necessity of cache on Linux, where data
            // fetching is faster
            // Cache portion of the remote process's address space.
            // Fetching data over the socket connection to dbx is slow.
            // Might be faster if we were using a binary protocol to talk to
            // dbx, but would have to test. For now, this cache works best
            // if it covers the entire heap of the remote process. FIXME: at

At least on linux the cache is definitely needed. I turned it off and a heap dump took 9x longer. Also I think covering the entire heap is overkill, and I doubt was ever being done given how small the cache is. So I think this comment can just be removed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292995](https://bugs.openjdk.org/browse/JDK-8292995): improve the SA page cache

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [9f528186](https://git.openjdk.org/jdk/pull/10069/files/9f5281866c581f231102934402865d8d231a58f2)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [9f528186](https://git.openjdk.org/jdk/pull/10069/files/9f5281866c581f231102934402865d8d231a58f2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10069/head:pull/10069` \
`$ git checkout pull/10069`

Update a local copy of the PR: \
`$ git checkout pull/10069` \
`$ git pull https://git.openjdk.org/jdk pull/10069/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10069`

View PR using the GUI difftool: \
`$ git pr show -t 10069`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10069.diff">https://git.openjdk.org/jdk/pull/10069.diff</a>

</details>

### CODE REVIEW

**Review by kevinjwalls — APPROVED** [review]
Looks good.  
Interesting to see the old comments talking about when it would drive a dbx process and issue commands to inspect memory.

(Based on that I was just checking WindbgDebuggerLocal as that's maybe confusingly named - just noting here that it doesn't drive the "WinDbg" tool exactly, it uses the Debugger Engine from dbgeng.dll (which is good).)

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/windbg/WindbgDebuggerLocal.java [hunk 1]
```diff
@@ -120,19 +120,8 @@ public void checkAlignment(long address, long alignment) {
...
-      // Fetching data over the socket connection to dbx is slow.
-      // Might be faster if we were using a binary protocol to talk to
-      // dbx, but would have to test. For now, this cache works best
-      // if it covers the entire heap of the remote process. FIXME: at
-      // least should make this tunable from the outside, i.e., via
-      // the UI. This is a cache of 4096 4K pages, or 16 MB. The page
-      // size must be adjusted to be the hardware's page size.
-      // (FIXME: should pick this up from the debugger.)
-      initCache(4096, 4096);
+        initCache(4096, parseCacheNumPagesProperty(1024 * 64));
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/windbg/WindbgDebuggerLocal.java:123]: Not a review. Just a minor comment.
The indent does not match the above lines.
As I see, the indent in the updated files is not unified. There is even indent 3.

**Review by sspitsyn — APPROVED** [review]
Nice simplification. Looks good.
Thanks,
Serguei

**Comment by plummercj** [comment]
Thanks Kevin and Serguei!

/integrate

**Pull request closed without merging**